### PR TITLE
Add analysis-test test framework

### PR DESCRIPTION
**Pull request by c-parsons**
This framework allows for easy creation of unittest-like tests to make assertions
on the provider-values returned by real targets.

Extensive documentation (in the form of updating the repository README) to follow in a future PR

### CODE REVIEW

**Comment by c-parsons** [comment]
Hello reviewers,

I propose deferral of writing a more-extensive README for the new framework until a future PR. I wanted to ensure the new API looked good before I went ahead and wrote a full document.
Please let me know if you'd prefer I not wait.

File: lib/unittest.bzl [hunk 1]
```diff
@@ -32,6 +32,15 @@ def register_unittest_toolchains():
...
 TOOLCHAIN_TYPE = "@bazel_skylib//toolchains/unittest:toolchain_type"
 
+# A private attribute used to denote a test rule is an analysis test.
+# For backwards compatibility, "unit tests" are not analysis tests, as being
+# an analysis test enforces the following test restrictions:
+#
+# 1. Test implementation functions may not register their own adhoc actions.
+# 2. There is a hard limit on the number of allowed transitive dependencies of
+#    analysis-test targets.
+_IS_ANALYSIS_TEST_ATTR_NAME = "_is_analysis_test"
```
Comment (thomasvl) [inline lib/unittest.bzl]: Is this some constant that also exists in bazel and has to match it?

File: lib/unittest.bzl [hunk 2]
```diff
@@ -59,6 +68,19 @@ unittest_toolchain = rule(
     },
 )
 
+def _impl_function_name(impl):
+    """Derives the name of the given rule implementation function.
+
+    This can be used for better test feedback."""
```
Comment (thomasvl) [inline lib/unittest.bzl]: Normally the closing '"""' goes on the next line.

File: lib/unittest.bzl [hunk 3]
```diff
@@ -196,22 +273,27 @@ def _end(env):
...
-    tc = env.ctx.toolchains[TOOLCHAIN_TYPE].unittest_toolchain_info
-    testbin = env.ctx.actions.declare_file(env.ctx.label.name + tc.file_ext)
-    if env.failures:
-        cmd = tc.failure_templ % tc.join_on.join(env.failures)
+    if getattr(env.ctx.attr, _IS_ANALYSIS_TEST_ATTR_NAME, False):
+        return [AnalysisTestResultInfo(
+            success = (len(env.failures) == 0),
+            message = "\n".join(env.failures),
+        )]
     else:
```
Comment (thomasvl) [inline lib/unittest.bzl]: Given the above returns, maybe drop the `else:` and drop the indent on the rest of this?

File: lib/unittest.bzl [hunk 4]
```diff
@@ -308,7 +390,22 @@ def _assert_new_set_equals(env, expected, actual, msg = None):
             full_msg = expectation_msg
         _fail(env, full_msg)
 
+def _expect_failure(env, expected_failure_msg = ""):
```
Comment (thomasvl) [inline lib/unittest.bzl]: Missing doc string since it is a public method?

File: lib/unittest.bzl [hunk 5]
```diff
@@ -196,22 +273,27 @@ def _end(env):
     that the results are reported.
 
     Args:
-      env: The test environment returned by `unittest.begin`.
+      env: The test environment returned by `unittest.begin` or `analysistest.begin`.
```
Comment (thomasvl) [inline lib/unittest.bzl]: Give the impl here forks with no overlap between analysis and unit, why both with the private attribute and conditional?  i.e. - have _end_unittest & _end_analysis, and set them into the structs independently.  You could still use the private attribute to ensure the correct begin/end are being called. As done, it actually allows the wrong begin/end to be called, if you do want to support that, then I'd suggest documenting that as an explicit impl detail.

File: tests/unittest_tests.bzl [hunk 6]
```diff
@@ -10,12 +26,146 @@ def _basic_passing_test(ctx):
...
+    change_setting_fake_rule(
+        name = "change_setting_fake_target",
+        tags = ["manual"],
+    )
+
+    failure_testing_test(
+        name = "failure_testing_test",
+        target_under_test = ":failure_testing_fake_target",
+    )
+    failure_testing_fake_rule(
```
Comment (thomasvl) [inline tests/unittest_tests.bzl]: add blank line for consistency?

File: lib/unittest.bzl (hunk 1 above)
Comment (c-parsons) [reply lib/unittest.bzl]: It is not. I just wanted to pull a constant out within this file and explain its significance

File: tests/unittest_tests.bzl (hunk 6 above)
Comment (c-parsons) [reply tests/unittest_tests.bzl]: The intention was to separate them into test-units (the test and the target it requires). Do you think it's worse this way?

File: lib/unittest.bzl (hunk 4 above)
Comment (c-parsons) [reply lib/unittest.bzl]: Done

File: lib/unittest.bzl (hunk 5 above)
Comment (c-parsons) [reply lib/unittest.bzl]: You're absolutely right. There was more initial overlap, and then I cleaned it up, and then the union no longer makes sense.

File: lib/unittest.bzl (hunk 2 above)
Comment (c-parsons) [reply lib/unittest.bzl]: Done

File: lib/unittest.bzl (hunk 1 above)
Comment (c-parsons) [reply lib/unittest.bzl]: (But note now this constant is no longer needed)

File: tests/unittest_tests.bzl (hunk 6 above)
Comment (thomasvl) [reply tests/unittest_tests.bzl]: Ahh missed that.

File: lib/unittest.bzl [hunk 7]
```diff
@@ -308,7 +389,31 @@ def _assert_new_set_equals(env, expected, actual, msg = None):
...
+
+    Args:
+      env: The test environment returned by `unittest.begin`.
+      expected_failure_msg: The error message to expect as a result of analysis failures.
+    """
+    dep = getattr(env.ctx.attr, "target_under_test")[0]
+    if AnalysisFailureInfo in dep:
+        dep_failure = dep[AnalysisFailureInfo]
+        actual_errors = ""
+        for cause in dep[AnalysisFailureInfo].causes.to_list():
```
Comment (dkelmer) [inline lib/unittest.bzl]: And the award for most helpful PR comment goes to.....

There are fancy ways for doing this. But having said that, I'm not sure if they're more efficient/idiomatic. Nor what they are. (but I think it uses join!).

File: lib/unittest.bzl [hunk 8]
```diff
@@ -308,7 +389,31 @@ def _assert_new_set_equals(env, expected, actual, msg = None):
...
+            actual_errors += cause.message + "\n"
+        if actual_errors.find(expected_failure_msg) < 0:
+            expectation_msg = "Expected errors to contain '%s' but did not. " % expected_failure_msg
+            expectation_msg += "Actual error:%s" % actual_errors
```
Comment (dkelmer) [inline lib/unittest.bzl]: Actual errrorS. If there was more than one they'd all get listed

File: tests/unittest_test.sh [hunk 9]
```diff
@@ -94,4 +104,11 @@ function test_basic_failing_test() {
   expect_log "In test _basic_failing_test from //tests:unittest_tests.bzl: Expected \"1\", but got \"2\""
 }
 
+function test_fail_unexpected_passing_test() {
+  ! bazel test //testdir:fail_unexpected_passing_test --test_output=all --verbose_failures \
```
Comment (dkelmer) [inline tests/unittest_test.sh]: Is the exclamation mark intentional? What does it do?

File: tests/unittest_tests.bzl [hunk 10]
```diff
@@ -1,5 +1,23 @@
-load("//lib:unittest.bzl", "asserts", "unittest")
+load("//lib:unittest.bzl", "analysistest", "asserts", "unittest")
 
+"""Unit tests for unittest.bzl."""
+
+###################################
+####### fail_basic_test ########
```
Comment (dkelmer) [inline tests/unittest_tests.bzl]:
```suggestion
####### fail_basic_test ###########
```

File: tests/unittest_tests.bzl [hunk 11]
```diff
@@ -10,12 +28,146 @@ def _basic_passing_test(ctx):
...
+    _change_setting_test,
+    config_settings = {
+        "//command_line_option:minimum_os_version": "1234.5678",
+    },
+)
+
+####################################
+####### failure_testing_test #######
+####################################
+def _failure_testing_test(ctx):
```
Comment (dkelmer) [inline tests/unittest_tests.bzl]: What are other examples of legit failures one might test for?

File: lib/unittest.bzl (hunk 7 above)
Comment (c-parsons) [reply lib/unittest.bzl]: a basic loop is fine with me if you don't feel strongly otherwise ;)

File: tests/unittest_test.sh (hunk 9 above)
Comment (c-parsons) [reply tests/unittest_test.sh]: It is! It is for testing intentional failures. exclamation mark is a negation of return code.
there are two types of tests.
1.  bazel test <target> || fail "should've passed"    (the fail only occurs if bazel test failed)
2. ! bazel test <target> || fail "should've failed" (the fail only occurs if the bazel test succeded)

File: tests/unittest_tests.bzl (hunk 11 above)
Comment (c-parsons) [reply tests/unittest_tests.bzl]: Well, you can theoretically test for any sort of error (including an index out of bounds exception), but I think those would be pretty terrible practice. I think the only kind of error we want to encourage expecting is one caused by fail().

File: tests/unittest_tests.bzl (hunk 10 above)
Comment (c-parsons) [reply tests/unittest_tests.bzl]: Done!

File: lib/unittest.bzl (hunk 8 above)
Comment (c-parsons) [reply lib/unittest.bzl]: fair point! updated

**Pull request merged**